File: basic_dsa_one_hundred_sixty_eight.java
/**
 * Basic DSA Problems - Validate Binary Search Tree
 * Check if tree is a valid BST (left < root < right, recursively)
 */
public class basic_dsa_one_hundred_sixty_eight {

    // Problem: Given root, return true if it is a valid BST.

    public static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    /**
     * Pass (min, max) allowed range; root must be in (min,max), left in (min, root), right in (root, max).
     * Time O(n), Space O(h)
     */
    public static boolean isValidBST(TreeNode root) {
        return valid(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    static boolean valid(TreeNode node, long min, long max) {
        if (node == null) return true;
        if (node.val <= min || node.val >= max) return false;
        return valid(node.left, min, node.val) && valid(node.right, node.val, max);
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Validate BST");
        System.out.println("=".repeat(60));

        TreeNode root = new TreeNode(2);
        root.left = new TreeNode(1);
        root.right = new TreeNode(3);
        System.out.println("Valid BST -> " + isValidBST(root));
    }
}

File: basic_dsa_one_hundred_sixty_eight.py
"""
Basic DSA - Sqrt(x)
Given a non-negative integer x, compute and return the integer square root of x.
The return value should be the floor of the true square root.
"""


def my_sqrt(x):
    """
    Binary search on [0, x].
    Time Complexity: O(log x)
    Space Complexity: O(1)
    """
    if x < 2:
        return x
    lo, hi = 1, x // 2
    ans = 1
    while lo <= hi:
        mid = (lo + hi) // 2
        sq = mid * mid
        if sq == x:
            return mid
        if sq < x:
            ans = mid
            lo = mid + 1
        else:
            hi = mid - 1
    return ans


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Sqrt(x)")
    print("=" * 60)

    print(my_sqrt(4))  # 2
    print(my_sqrt(8))  # 2
